File: crm/app/internal/ports/party_repository.go
// Package ports defines inbound and outbound interfaces for the CRM hexagonal architecture.
// Nothing in this package may import sqlite, http, or any adapter package.
package ports

import (
	"context"

	"github.com/vantt/data-integration/crm/app/internal/domain"
)

// PartyRepository is the outbound port for persisting and querying Party golden records
// and their associated identities.
type PartyRepository interface {
	// FindByIdentity returns the party that owns the given identity, or nil if not found.
	FindByIdentity(ctx context.Context, identityType, identityValue string) (*domain.Party, error)

	// Create inserts a new party row.
	Create(ctx context.Context, p *domain.Party) error

	// GetByID returns a party by primary key, or nil if not found.
	GetByID(ctx context.Context, partyID string) (*domain.Party, error)

	// Update persists changes to mutable party fields (display_name, primary_phone,
	// primary_email, status, is_merged, merged_into).
	Update(ctx context.Context, p *domain.Party) error

	// ListByPhone returns all non-merged parties with a matching normalised primary_phone.
	ListByPhone(ctx context.Context, phone string) ([]domain.Party, error)

	// ListByEmail returns all non-merged parties with a matching normalised primary_email.
	ListByEmail(ctx context.Context, email string) ([]domain.Party, error)

	// SearchByName runs an FTS5 MATCH on crm_party_fts and returns matching party IDs.
	// The caller is responsible for post-filtering (e.g. phone-prefix narrowing).
	SearchByName(ctx context.Context, query string) ([]string, error)

	// UpsertIdentity inserts or ignores a party_identity row.
	// The UNIQUE(identity_type, identity_value) constraint is the dedup guarantee.
	UpsertIdentity(ctx context.Context, id *domain.PartyIdentity) error

	// ListIdentities returns all identity rows for a given party.
	ListIdentities(ctx context.Context, partyID string) ([]domain.PartyIdentity, error)

	// CreateWithIdentities inserts a party and all supplied identities in a single
	// atomic transaction. Prevents an orphan party row when identity attachment fails mid-way.
	CreateWithIdentities(ctx context.Context, p *domain.Party, identities []*domain.PartyIdentity) error

	// UpdateContactQuality updates the mutable contact_quality field on an identity row.
	// Valid values: masked | unverified | verified.
	UpdateContactQuality(ctx context.Context, identityID, quality string) error
}

File: crm/src/domain/ports/party_repository.py
from __future__ import annotations

from typing import Protocol

from domain.entities.party import Party, PartyIdentity


class PartyRepository(Protocol):
    """Outbound port for persisting and querying Party golden records and their identities."""

    def find_by_identity(
        self, identity_type: str, identity_value: str
    ) -> Party | None:
        """Return the party that owns the given identity, or None if not found."""
        ...

    def create(self, party: Party) -> Party:
        """Insert a new party row and return it (with generated party_id)."""
        ...

    def get_by_id(self, party_id: str) -> Party | None:
        """Return a party by primary key, or None if not found."""
        ...

    def update(self, party: Party) -> None:
        """Persist changes to mutable party fields (display_name, primary_phone,
        primary_email, status, is_merged, merged_into)."""
        ...

    def list_by_phone(self, phone: str) -> list[Party]:
        """Return all non-merged parties with a matching normalised primary_phone."""
        ...

    def list_by_email(self, email: str) -> list[Party]:
        """Return all non-merged parties with a matching normalised primary_email."""
        ...

    def search_by_name(self, query: str) -> list[str]:
        """Run FTS MATCH on crm_party_fts and return matching party IDs."""
        ...

    def upsert_identity(self, identity: PartyIdentity) -> None:
        """Insert or ignore a party_identity row.
        UNIQUE(identity_type, identity_value) is the dedup guarantee.
        source_contact_quality is immutable — INSERT OR IGNORE semantics.
        contact_quality is mutable — UPDATE on conflict."""
        ...

    def list_identities(self, party_id: str) -> list[PartyIdentity]:
        """Return all identity rows for a given party."""
        ...

    def create_with_identities(
        self, party: Party, identities: list[PartyIdentity]
    ) -> Party:
        """Insert a party and all supplied identities in a single atomic transaction.
        Prevents an orphan party row when identity attachment fails mid-way."""
        ...

    def update_contact_quality(self, identity_id: str, contact_quality: str) -> None:
        """Update the mutable contact_quality field on an identity row.
        Valid values: masked | unverified | verified."""
        ...
